plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+    void subscribe(const char * channel, StringAttr & value);
+    void unsubscribe(const char * channel);
+    void attachLibev();
+    bool lock(const char * key, const char * channel);
+    void handleLoop(struct ev_loop * evLoop, ev_tstamp timeout);
+
+    //callbacks
+    static void assertCallbackError(const redisAsyncContext * context, const redisReply * reply, const char * _msg);
+    static void connectCB(const redisAsyncContext *c, int status);
+    static void disconnectCB(const redisAsyncContext *c, int status);
+    static void subCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void pubCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void setCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void getCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void unsubCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void setLockCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void selectCB(redisAsyncContext * context, void * _reply, void * privdata);
+    static void timeoutCB(struct ev_loop * evLoop, ev_timer *w, int revents);
+
+protected :
+    redisAsyncContext * context;
+};
+
+class SubContainer : public AsyncConnection
+{
+public :
+    SubContainer(ICodeContext * ctx, const char * options, const char * _channel, unsigned __int64 _database);
+    SubContainer(ICodeContext * ctx, const char * options, const char * _channel, redisCallbackFn * _callback, unsigned __int64 _database);
+    ~SubContainer()
+    {
+        unsubscribe();
+    }
+
+    inline void setCB(redisCallbackFn * _callback) { callback = _callback; }
+    inline redisCallbackFn * getCB() { return callback; }
+    inline void setChannel(const char * _channel) { channel.set(_channel); }
+    inline const char * getChannel() { return channel.str(); }
+    inline void setMessage(const char * _message) { message.set(_message); }
+    inline const char * getMessage() { return message.str(); }
+    inline void wait(unsigned timeout) { msgSem.wait(timeout); }
+    inline void signal() { msgSem.signal(); }
+    inline void subActivated() { subActiveSem.signal(); }
+    inline void subActivationWait(unsigned timeout) { subActiveSem.wait(timeout); }
+    void subscribe(struct ev_loop * evLoop);
+    void unsubscribe();
+
+    void stopEvLoop()
+    {
+        CriticalBlock block(crit);
+        context->ev.delRead((void*)context->ev.data);
+        context->ev.delWrite((void*)context->ev.data);
+    }
+    //callback
+    static void subCB(redisAsyncContext * context, void * _reply, void * privdata);
+
+protected :
+    Semaphore msgSem;
+    Semaphore subActiveSem;
+    redisCallbackFn * callback;
+    StringAttr message;
+    StringAttr channel;
+};
+
+class SubscriptionThread : implements IThreaded, implements IInterface, public SubContainer
```

Comment:
Originally I wasn't sure I would want to reuse an instance of this class with another callback/redis-operator. I tried to keep a lot of this code (the plugin in general) generic and open as it is far from complete, regarding the incorporation of a lot of the really useful redis commands. Pipelining commands, sets, lists, and allowing for operations server-side such as sorting.